Track submissions that would be rate limited

dannyroberts opened the pull request:
##### SUMMARY

Get the system in place for rate limiting submissions, but for now
instead of actually blocking usage, just report to datadog that a
submission would have been blocked.

The rate limiting is based on the domain-wide per-user rate, and is
given both a minimum usage to allow (e.g. for a domain with very few
users) and a per user rate for projects that have enough users.

dannyroberts committed:
Track submissions that would be rate limited

Get the system in place for rate limiting submissions, but for now
instead of actually blocking usage, just report to datadog that a
submission would have been blocked.

The rate limiting is based on the domain-wide per-user rate, and is
given both a minimum usage to allow (e.g. for a domain with very few
users) and a per user rate for projects that have enough users.

dannyroberts committed:
Fix stickler warnings

snopoke commented on corehq/apps/receiverwrapper/rate_limiter.py:
```diff
@@ -0,0 +1,30 @@
+)
+
+test_rates = PerUserRateDefinition(
+    per_user_rate_definition=rates_promised_not_to_go_lower_than.times(2.0),
```
why `times(2.0)`?

dannyroberts replied:
I wanted to try an approach of starting out high and rate limiting only extreme outliers before pulling it in and landing somewhere comfortable. In order to get a good sense of where to end up I wanted to start at the high end of the range. What I don't want to do is to kind of slam on new restrictions to our users and have everyone scramble to adjust to them. (I realize this PR doesn't actually rate limit yet, but I wanted to collect data for the place I was considering starting rate limiting at.)

Does that approach make sense?

snopoke commented on corehq/apps/receiverwrapper/rate_limiter.py:
```diff
@@ -0,0 +1,30 @@
+    per_day=23,
+    per_hour=3,
+    per_minute=0.07,
+    per_second=0.005,
```
I don't understand how this makes sense. How can your 1s counter make sense of this value? I don't think you can rate limit below the 1/s.

dannyroberts replied:
Yeah so this is what the `floor_for_small_domain` is about. The intention here is that (using `rates_promised_not_to_go_lower_than` and `floor_for_small_domain ` as an example, even though we're starting out with using `rates_promised_not_to_go_lower_than.times(2)`) the total effective limits of a domain would follow the table below for different numbers of users

| # of users | second | minute | hour | day | week
| --- | --- | --- | --- | --- | --- |
| 1 | 1 | 10 | 30 | 50 | 115 |
| 10 | 1 | 10 | 30 | 230 | 1150 |
| 100 | 1 | 7 | 300 | 2300 | 11500 |
| 1000 | 5 | 70 | 3000 | 23000 | 115000 |
| 10000 | 50 | 700 | 30000 | 230000 | 1150000 |
| 100000 | 500 | 7000 | 300000 | 2300000 | 11500000 |

Notice in particular how for low numbers it is more like `floor_for_small_domain`, and for high numbers of users, `floor_for_small_domain` becomes irrelevant and it just grows linearly according to `rates_promised_not_to_go_lower_than`.

snopoke replied:
OK, that's what I thought but it still didn't make a lot of sense except purely for 'documentation' purposes. Fine to leave it though.

snopoke commented on corehq/apps/receiverwrapper/views.py:
```diff
@@ -64,6 +65,11 @@
+        datadog_counter('commcare.xform_submissions.rate_limited.test', tags=[
+            'domain:{}'.format(domain),
+        ])
+    submission_rate_limiter.report_usage(domain)
```
can we ignore this block if `settings.ENTERPRISE_MODE`

dannyroberts replied:
Yeah makes total sense. In addition, once this gets a little further along, I am planning on pulling the specific rates out into localsettings or otherwise make it environment-specific. There's certainly no reason other environments would want to use the same rate-limiting figures as production

dannyroberts committed:
Restrict rate limit testing to non ENTERPRISE_MODE only

dannyroberts committed:
Report any errors with new rate limiting code without 500ing

While I expect the rate limiting code not to error hard
(it gets called in tests and they passed)
I want any subtler issues to get raised without
interrupting submissions.
This does not help with any potential performance issues,
but I am less concerned about those since on the vast majority of requests
this should only add a few requests to redis.

In testing, errors will still bubble up to produce test errors.
